Export new metrics and add status label

ibizaman opened the pull request:
This PR does two things. I can split them if you want.

First, it adds a state label to torrent specific metrics. I wanted to only show torrents that are in downloading state on some graphs and adding the label ok the metrics was the easiest. I tried first using a query join transformation in grafana and other transforms but couldn’t figure it out.

Second, it adds a few metrics coming from deluge like the added one that was requested in another issue.

Here’s a snapshot of what I’m using this for.
![570B01EA-9875-4CF3-9ADA-3C75973D67AC](https://user-images.githubusercontent.com/1044950/150728319-a8ff7e95-36a3-4a2e-9cbe-80347de74e49.jpeg)

And here’s the dashboard definition https://gist.github.com/ibizaman/e4a79c2e1800a74c25ba3755aa75426a

ibizaman committed:
add state label to torrent metrics

ibizaman committed:
add more per-torrent metrics

tobbez commented on deluge_exporter.py:
```diff
@@ -136,10 +136,12 @@ def collect(self):
         }
         torrents_by_label = defaultdict(int)
-        for torrent in client.core.get_torrents_status({}, [b"label", b"state", b"download_payload_rate", b"upload_payload_rate"]).values():
+        torrents_with_state = defaultdict(int)
```
Rename to torrent_states and change type to regular dict (defaultdict is not needed here). That makes it:

```python
torrent_states = {}
```

tobbez commented on deluge_exporter.py:
```diff
@@ -136,10 +136,12 @@ def collect(self):
         torrents_by_label = defaultdict(int)
-        for torrent in client.core.get_torrents_status({}, [b"label", b"state", b"download_payload_rate", b"upload_payload_rate"]).values():
+        torrents_with_state = defaultdict(int)
+        for torrent in client.core.get_torrents_status({}, [b"name", b"label", b"state", b"download_payload_rate", b"upload_payload_rate"]).values():
```
The states dict should be keyed by info hash, since names are not guaranteed to be unique, meaning this becomes:

```py
for torrent_hash, torrent in client.core.get_torrents_status({}, [b"label", b"state", b"download_payload_rate", b"upload_payload_rate"]).items():
```

tobbez commented on deluge_exporter.py:
```diff
@@ -136,10 +136,12 @@ def collect(self):
             if b"label" in torrent:
                 torrents_by_label[torrent[b"label"].decode("utf-8")] += 1
             torrents_by_state[torrent[b"state"].decode("utf-8").lower()] += 1
+            torrents_with_state[torrent[b"name"]] = torrent[b"state"].decode("utf-8").lower()
```
```py
torrent_states[torrent_hash] = torrent[b"state"].decode("utf-8").lower()
```

(as previously: rename the dict and use the hash as the key)

tobbez commented on deluge_exporter.py:
```diff
@@ -164,12 +166,24 @@ def collect(self):
                 (GaugeMetricFamily, b"total_seeds", "The number of seeds in the swarm for this torrent"),
+                (GaugeMetricFamily, b"active_time", "The number of seconds this torrent has been active. i.e. not paused"),
+                (GaugeMetricFamily, b"seeding_time", "The number of seconds this torrent has been active and seeding"),
+                (GaugeMetricFamily, b"finished_time", "When this torrent finished being active"),
```
Incorrect. This is `The number of seconds this torrent has spent in the finished state`.

tobbez commented on deluge_exporter.py:
```diff
@@ -164,12 +166,24 @@ def collect(self):
+                (GaugeMetricFamily, b"active_time", "The number of seconds this torrent has been active. i.e. not paused"),
+                (GaugeMetricFamily, b"seeding_time", "The number of seconds this torrent has been active and seeding"),
+                (GaugeMetricFamily, b"finished_time", "When this torrent finished being active"),
+                (GaugeMetricFamily, b"all_time_download", "The amount of time downloading the torrent"),
```
Incorrect. This is the total number of payload bytes downloaded across all sessions.

tobbez commented on deluge_exporter.py:
```diff
@@ -164,12 +166,24 @@ def collect(self):
+                (GaugeMetricFamily, b"time_since_download", "The number of seconds since we last downloaded payload from a peer on this torrent"),
+                (GaugeMetricFamily, b"time_since_upload", "The number of seconds since we last uploaded payload to a peer on this torrent"),
+                (GaugeMetricFamily, b"time_since_transfer", "The number of seconds since we last uploaded payload to or from a peer on this torrent"),
+                (GaugeMetricFamily, b"last_seen_complete", "When was the last time we saw the complete torrent on the swarm"),
```
Better expressed as "The time when we last saw a seed or peers that together formed a complete copy of the torrent".

tobbez commented on deluge_exporter.py:
```diff
@@ -164,12 +166,24 @@ def collect(self):
+                (GaugeMetricFamily, b"time_since_upload", "The number of seconds since we last uploaded payload to a peer on this torrent"),
+                (GaugeMetricFamily, b"time_since_transfer", "The number of seconds since we last uploaded payload to or from a peer on this torrent"),
+                (GaugeMetricFamily, b"last_seen_complete", "When was the last time we saw the complete torrent on the swarm"),
+                (GaugeMetricFamily, b"progress", "Torrent progress percentage"),
```
I'm not convinced this should be included. It's a derived metric that can be calculated using `total_done/total_size`.

tobbez commented on deluge_exporter.py:
```diff
@@ -164,12 +166,24 @@ def collect(self):
+                (GaugeMetricFamily, b"time_since_transfer", "The number of seconds since we last uploaded payload to or from a peer on this torrent"),
+                (GaugeMetricFamily, b"last_seen_complete", "When was the last time we saw the complete torrent on the swarm"),
+                (GaugeMetricFamily, b"progress", "Torrent progress percentage"),
+                (GaugeMetricFamily, b"seeds_peers_ratio", "Torrent ratio"),
```
Incorrect. This is the ratio between the number of seeds and number of peers, which should already be possible to calculate (using num_peers and num_seeds). It can thus be removed.

The actual torrent ratio can also be calculated using other metrics (`total_uploaded/all_time_download`).

tobbez commented on deluge_exporter.py:
```diff
@@ -164,12 +166,24 @@ def collect(self):
             for torrent_hash, torrent in client.core.get_torrents_status({}, [key[1] for key in per_torrent_keys] + [b"name"]).items():
                 for metric_name, metric in per_torrent_metrics.items():
-                    metric.add_metric([torrent[b"name"].decode("utf-8"), torrent_hash.decode("utf-8")], torrent[metric_name])
+                    metric.add_metric([torrent[b"name"].decode("utf-8"), torrent_hash.decode("utf-8"), torrents_with_state[torrent[b"name"]]], torrent[metric_name])
```
```python
metric.add_metric([torrent[b"name"].decode("utf-8"), torrent_hash.decode("utf-8"), torrent_states[torrent_hash]], torrent[metric_name])
```

(as previously: rename the dict and use the hash as the key)

tobbez reviewed (changes requested):
Using the same PR is acceptable (assuming unrelated changes are separated into different commits), but splitting it up would have been better.

The changes looks fine apart from the comments inline.

While you're rewriting the commit history, please also update the message of the "add more metrics" commit to "add more per-torrent metrics".

It's a bit unfortunate that Deluge inconsistently renames some of the libtorrent metrics (e.g. libtorrent's `all_time_upload` is renamed to `total_uploaded` while `all_time_download` preserves the libtorrent name). We'll just have to live with that for now (the fix would be to add functionality to Deluge to return the raw per-torrent libtorrent metrics).

Some of the metrics added are deprecated in libtorrent 1.2 and removed in 2.0 (`time_since_upload`, `time_since_download`, `active_time`, `finished_time`, `seeding_time`), but I suppose we can keep them until Deluge migrates to libtorrent 2.0.

ibizaman commented:
Thank you for the thorough review! I implemented all your comments. I also added the metrics to the Readme as well as adding `state=` as label and also fixed the naming of the `deluge_torrent_total_*` metrics.

tobbez commented on deluge_exporter.py:
```diff
@@ -164,12 +166,22 @@ def collect(self):
                 (GaugeMetricFamily, b"num_seeds", "The number of seeds currently connected to for this torrent"),
                 (GaugeMetricFamily, b"total_peers", "The number of peers in the swarm for this torrent"),
                 (GaugeMetricFamily, b"total_seeds", "The number of seeds in the swarm for this torrent"),
+                (GaugeMetricFamily, b"active_time", "The number of seconds this torrent has been active. i.e. not paused")
```
Missing trailing comma.

tobbez commented on deluge_exporter.py:
```diff
@@ -136,10 +136,12 @@ def collect(self):
         torrents_by_label = defaultdict(int)
-        for torrent in client.core.get_torrents_status({}, [b"label", b"state", b"download_payload_rate", b"upload_payload_rate"]).values():
+        torrents_states = {}
+        for torrent_hash, torrent in client.core.get_torrents_status({}, [b"label", b"state", b"download_payload_rate", b"upload_payload_rate"]).values():
```
This needs to be `.items()`, not `.values()`.

tobbez reviewed (changes requested):
Two issues that need to be fixed (the current version won't work at all when per-torrent metrics are enabled).

You can also skip updating the metrics table if you like. I need to regenerate it after this PR is merged anyway.

ibizaman commented:
Sorry for the wait. I didn't update the metrics table as you suggested.

This I tested it on my local server. Hopefully I didn’t miss anything.
